### PR TITLE
feat: upgrade to babel 7

### LINKED ISSUE #26: Incentivise to just compile at the "app" level
See: https://github.com/moxystudio/react-with-moxy/pull/99
Also: https://github.com/parcel-bundler/parcel/pull/559#discussion_r161926651

Important note: re-enable `babel-plugin-transform-react-remove-prop-types` and `babel-plugin-transform-react-inline-elements` when done.

### PR DESCRIPTION
**Pull request by satazor**
BREAKING CHANGE: the way the preset is configured is different

Closes #26

### CODE REVIEW

**Comment by satazor** [comment]
@Areadrill I took some inspiration on https://github.com/moxystudio/babel-preset-moxy/tree/script-division but I've simplified the folder structure setup. Essentially, it stayed the same as it is now. Having that said, we don't need a diagram to help people out, but thanks for taking the time into drawing it!

File: README.md [hunk 1]
```diff
@@ -21,89 +21,151 @@
...
+
+## Do I need to transpile?
+
+There has been [discussion](https://github.com/parcel-bundler/parcel/pull/559#discussion_r161926651) in the community about libraries not being compiled, leaving all compilation efforts to top-level projects consuming them. This makes sense, since developers know what platforms their top-level project target and are able to compile their dependencies accordingly. Furthermore, library maintainers are constantly having to update their compilation options as new features are accepted into different stages of the specification, which creates significant and unnecessary overhead.
+
+Problems arise, however, in libraries which target both Node.js and browser, or if non-standard JavaScript is being used, such as [proposals](https://github.com/tc39/proposals) or [JSX](https://reactjs.org/docs/introducing-jsx.html). In those situations, library authors are required to transpile their libraries' code to offer `CommonJS` and `ES` module variants or to transform non-standard JavaScript to standard JavaScript.
```
Comment (paulobmarcos) [inline README.md]: `'`  after "their libraries"

File: lib/dynamic-import.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+const addDynamicImportSupport = (config, modules) => {
+    console.log('modules', modules);
```
Comment (acostalima) [inline lib/dynamic-import.js]: Left over?

File: README.md [hunk 3]
```diff
@@ -21,89 +21,151 @@
...
 
-Create `.babelrc` at the root of your project:
+### 1. Choose a preset-type
 
-```json
-{
-    "presets": ["moxy"]
-}
-```
+There's two preset types available for you to use:
```
[the remark this answers is not in the document]
Comment (satazor) [reply README.md]: Oops, please amend it 👍

File: lib/dynamic-import.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (satazor) [reply lib/dynamic-import.js]: Yep

**Comment by acostalima** [comment]
Well done! 👏

File: lib.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    );
+
+    // Adds dynamic import support
+    if (options.dynamicImport) {
+        addDynamicImportSupport(config, options.modules);
+    }
+
+    // Add react support without doing any development or production transformations
+    if (options.react) {
+        addReactSupport(config);
```
Comment (satazor) [inline lib.js]: Maybe it’s worth to explicitly pass null in the second argument for readability

File: end-project.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        // Set Nodejs target to be the latest LTS
+        targets: Array.isArray(options.targets) ? Object.assign({},
+            options.targets.indexOf('node') !== -1 ? { node: '8.9' } : {},
+            options.targets.indexOf('browsers') !== -1 ? { browsers: ['extends browserslist-config-google'] } : {}
+        ) : options.targets,
+        // Enables support for builtin/feature proposals that have native support by the defined target environments
+        shippedProposals: options.shippedProposals,
+    }]);
+
+    // The two plugins bellow activate stage 3 features that babel hasn't added to the stage 3 preset yet
```
Comment (satazor) [inline end-project.js]: `The two plugins` no longer make sense here. Can you adjust?

File: lib.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    if (moduleTransformations[options.modules]) {
+        config.plugins.push(`@babel/plugin-${moduleTransformations[options.modules]}`);
+    }
+
+    // The two plugins above activate stage 3 features that babel hasn't added to the stage 3 preset yet
+    config.plugins.push(
+        // Allows class { handleClick = () => { } static propTypes = { foo: PropTypes.string } }
+        require.resolve('@babel/plugin-proposal-class-properties'),
+        // Support destructuring of objects, e.g.: { ...foo }
+        [require.resolve('@babel/plugin-proposal-object-rest-spread'), { useBuiltIns: true }]
```
Comment (satazor) [inline lib.js]: Remove this plugin as it's part of ECMA9

File: lib.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        presets: [],
+        plugins: [],
+    };
+
+    // Activate modules transformation
+    if (moduleTransformations[options.modules]) {
+        config.plugins.push(`@babel/plugin-${moduleTransformations[options.modules]}`);
+    }
+
+    // The two plugins above activate stage 3 features that babel hasn't added to the stage 3 preset yet
```
Comment (satazor) [inline lib.js]: `The two plugins` no longer make sense here. Can you adjust?

File: package.json [hunk 8]
```diff
@@ -49,28 +49,31 @@
     ]
   },
   "dependencies": {
+    "@babel/plugin-proposal-class-properties": "^7.3.0",
```
Comment (satazor) [inline package.json]: We don't need this anymore.

Comment (paulobmarcos) [reply package.json]: Are you referring  to this one:  `@babel/plugin-proposal-object-rest-spread": "^7.3.2` ?

File: lib/react.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        [require.resolve('@babel/plugin-transform-react-jsx'), { useBuiltIns: true }],
+        require.resolve('@babel/plugin-transform-react-display-name')
+    );
+
+    // Enable optimizations on production
+    if (env === 'production') {
+        config.plugins.push(
+            [require.resolve('babel-plugin-transform-react-remove-prop-types'), { removeImport: true }]
+        );
+    // The following two plugins are currently necessary to make React warnings include more valuable information.
```
Comment (satazor) [inline lib/react.js]: Nifty pick, remove the `.` from the end of  both sentences.

File: test/end-project.spec.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    });
+
+    it('should allow overriding options.targets with an object', () => {
+        expect(preset(null, { targets: { node: 'current' } })).toMatchSnapshot();
+    });
+});
+
+describe('env', () => {
+    it('should take into consideration process.env.NODE_ENV', () => {
+        process.env.NODE_ENV = 'development';
```
Comment (satazor) [inline test/end-project.spec.js]: I forgot to reset the `process.env.NODE_ENV`. Could you add the following?

```js
const env = process.env.NODE_ENV;

afterEach(() => {
    process.env.NODE_ENV = env;
})
```

**Pull request merged**